pkg/imgpkg/imageset/image_set.go

```diff
@@ -77,57 +79,135 @@ func (o *ImageSet) Import(imgOrIndexes []imagedesc.ImageOrIndex,
+	for _, item := range imgOrIndexes {
+		item := item // copy
+
+		go func() {
+			importThrottle.Take()
+			defer importThrottle.Done()
+
+			o.verifyProcessedImages(item, errChVerifyImages, importRepo, registry, importedImages)
 		}()
 	}
 
+	err = checkForAnyAsyncErrors(imgOrIndexes, errChVerifyImages)
+	if err != nil {
+		return nil, err
+	}
+
+	return importedImages, nil
+}
+
+func checkForAnyAsyncErrors(imgOrIndexes []imagedesc.ImageOrIndex, errCh chan error) error {
 	for i := 0; i < len(imgOrIndexes); i++ {
 		err := <-errCh
 		if err != nil {
-			return nil, err
+			return err
 		}
 	}
+	return nil
+}
 
-	return importedImages, nil
+func addImageOrImageIndexToTaggableMapForMultiWrite(item imagedesc.ImageOrIndex, errCh chan error, importRepo regname.Repository, registry ImagesReaderWriter, imageOrIndexesToWrite map[regname.Reference]regremote.Taggable, lock *sync.Mutex) {
+	itemDigest, err := item.Digest()
+	if err != nil {
+		errCh <- err
+		return
+	}
+
+	tag := fmt.Sprintf("imgpkg-%s-%s", itemDigest.Algorithm, itemDigest.Hex)
+	uploadTagRef, err := regname.NewTag(fmt.Sprintf("%s:%s", importRepo.Name(), tag))
+	if err != nil {
+		errCh <- fmt.Errorf("Building upload tag image ref: %s", err)
+		return
+	}
+
+	switch {
+	case item.Image != nil:
+		itemRef, err := regname.ParseReference(item.Ref())
+		if err != nil {
+			errCh <- fmt.Errorf("Unable to parse reference: %s: %s", item.Ref(), err)
+			return
+		}
+
+		imageToWrite := regv1.Image(*item.Image)
+
+		if imageBlobsCanBeMounted(itemRef, uploadTagRef) {
+			imageToWrite, err = item.MountableImage(registry)
+			if err != nil {
+				imageToWrite = regv1.Image(*item.Image)
+			}
+		}
+
+		lock.Lock()
+		defer lock.Unlock()
+		imageOrIndexesToWrite[uploadTagRef] = imageToWrite
+	case item.Index != nil:
+		lock.Lock()
+		defer lock.Unlock()
+		imageOrIndexesToWrite[uploadTagRef] = *item.Index
+	}
+
+	errCh <- nil
+}
+
+func (o *ImageSet) verifyProcessedImages(item imagedesc.ImageOrIndex, errCh chan error, importRepo regname.Repository, registry ImagesReaderWriter, importedImages *ProcessedImages) {
+	existingRef, err := regname.NewDigest(item.Ref())
+	if err != nil {
+		errCh <- err
+		return
+	}
+
+	importDigestRef, err := o.tagAndVerifyImage(item, existingRef, importRepo, registry)
+	if err != nil {
+		errCh <- fmt.Errorf("Importing image %s: %s", existingRef.Name(), err)
+		return
+	}
+
+	var regImage regv1.Image
+	if item.Image != nil {
+		regImage = *item.Image
+	}
+	var regImageIndex regv1.ImageIndex
+	if item.Index != nil {
+		regImageIndex = *item.Index
+	}
+	importedImages.Add(ProcessedImage{
+		UnprocessedImageRef: UnprocessedImageRef{existingRef.Name(), item.Tag()},
+		DigestRef:           importDigestRef.Name(),
+		Image:               regImage,
+		ImageIndex:          regImageIndex,
+	})
+	errCh <- nil
 }
 
-func (o *ImageSet) importImage(item imagedesc.ImageOrIndex,
+func (o *ImageSet) tagAndVerifyImage(item imagedesc.ImageOrIndex,
```

Comment:
i think we should split this function into two: one that verifies things and one that adds additional tag (at glance dont see anything being reused between two parts).

Reply:
I can make this change, although it's not obvious to me why. It sounds like this function is doing too much?